src/runtime/vm/pooled_allocator.h

```diff
@@ -57,14 +57,22 @@ class PooledAllocator final : public Allocator {
     Buffer buf;
     buf.device = device_;
     buf.size = size;
-    buf.data = DeviceAPI::Get(device_)->AllocDataSpace(device_, size, alignment, type_hint);
+    try {
+      buf.data = DeviceAPI::Get(device_)->AllocDataSpace(device_, size, alignment, type_hint);
+    } catch (InternalError& err) {
+      LOG(WARNING) << "PooledAllocator got InternalError during allocation: " << err.message();
+      LOG(WARNING) << "Trying to release all unused memory and reallocate...";
+      ReleaseAll();
+      buf.data = DeviceAPI::Get(device_)->AllocDataSpace(device_, size, alignment, type_hint);
```

Comment:
What to expect if it still failed here?

Reply:
If it still fails, an InternalError will be thrown, causing a TVMError regarding OOM in the Python End.

Reply:
Hmm, would that be better if we let ReleaseAll return the size it released and check if it is larger than the requested size? So that we can directly throw a message including both sizes without calling alloc again.

Reply:
Thanks for the suggestion. But IMHO this is not robust enough.

Say that we have 8 GB GPU memory, the pooled cached 4 GB and we want to allocate 6 GB. 

- Applying your idea, `ReleaseAll()` returns "4GB" which is less than "6GB", thus resulting in a failed allocation.
- Instead, if we release unused memory and do re-allocation, "6GB" is very likely to be successfully allocated.

But the big picture behind your idea is practical if we can have some APIs like "total_system_memory" and "available_system_memory", which may require introducing a series of system driver libraries. e.g., `cudaMemGetInfo` by CudaRT (user space) or `NVML` (more system privilege is allowed).

Reply:
Fair enough. I don't have other comments then.